tests/unit/fem/test_linearform_ext.cpp

```diff
@@ -0,0 +1,152 @@
+// Copyright (c) 2010-2022, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#include "mfem.hpp"
+#include "unit_tests.hpp"
+#include <functional>
+
+using namespace mfem;
+
+struct LinearFormExtTest
+{
+   enum { DomainLF, DomainLFGrad, VectorDomainLF, VectorDomainLFGrad };
+   using vector_fct_t = void(const Vector&, Vector&);
+   const double abs_tol = 1e-14, rel_tol = 1e-14;
+   const char *mesh_filename;
+   Mesh mesh;
+   const int dim, vdim, ordering, gll, problem, p, q, SEED = 0x100001b3;
+   H1_FECollection fec;
+   FiniteElementSpace vfes, mfes;
+   const Geometry::Type geom_type;
+   IntegrationRules IntRulesGLL;
+   const IntegrationRule *irGLL, *ir;
+   Array<int> elem_marker;
+   Vector one_vec, dim_vec, vdim_vec, vdim_dim_vec;
+   ConstantCoefficient cst_coeff;
+   VectorConstantCoefficient dim_cst_coeff, vdim_cst_coeff, vdim_dim_cst_coeff;
+   std::function<vector_fct_t> vdim_vec_function =
+   [&](const Vector&, Vector &y) { y.SetSize(vdim); y.Randomize(SEED); };
+   std::function<vector_fct_t> vector_fct;
+   VectorFunctionCoefficient vdim_fct_coeff;
+   LinearForm lf_dev, lf_std;
+   LinearFormIntegrator *lfi_dev, *lfi_std;
+
+   LinearFormExtTest(const char *mesh_filename,
+                     int vdim, int ordering, int gll, int problem, int p):
+      mesh_filename(mesh_filename),
+      mesh(Mesh::LoadFromFile(mesh_filename)),
+      dim(mesh.Dimension()), vdim(vdim), ordering(ordering), gll(gll),
+      problem(problem), p(p), q(2*p + (gll?-1:3)), fec(p, dim),
+      vfes(&mesh, &fec, vdim, ordering), mfes(&mesh, &fec, dim),
+      geom_type(vfes.GetFE(0)->GetGeomType()),
+      IntRulesGLL(0, Quadrature1D::GaussLobatto),
+      irGLL(&IntRulesGLL.Get(geom_type, q)), ir(&IntRules.Get(geom_type, q)),
+      elem_marker(), one_vec(1), dim_vec(dim), vdim_vec(vdim), vdim_dim_vec(vdim*dim),
+      cst_coeff(M_PI), dim_cst_coeff((dim_vec.Randomize(SEED), dim_vec)),
+      vdim_cst_coeff((vdim_vec.Randomize(SEED), vdim_vec)),
+      vdim_dim_cst_coeff((vdim_dim_vec.Randomize(SEED), vdim_dim_vec)),
+      vector_fct(vdim_vec_function), vdim_fct_coeff(vdim, vector_fct),
+      lf_dev(&vfes), lf_std(&vfes), lfi_dev(nullptr), lfi_std(nullptr)
+   {
+      REQUIRE(mesh.attributes.Size() == 1);
+      for (int e = 0; e < mesh.GetNE(); e++) { mesh.SetAttribute(e, e%2?1:2); }
+      mesh.SetAttributes();
+      REQUIRE(mesh.attributes.Size() == 2);
```

Comment:
These two `REQUIRE`s aren't really necessary (I guess they don't hurt, but in particular the first one isn't a requirement for this test to work correctly)

Reply:
Removed first one, turned the second to MFEM_VERIFY.